crux-bench/time-series-bench/src/time_series_bench/ts_devices.clj

```diff
@@ -0,0 +1,300 @@
+                                                             (or [device-id :device-info/model "pinto"]
+                                                                 [device-id :device-info/model "focus"])]})
+                                           (reduce into []))
+                          db (crux/db node #inst "1970")]
+                      (with-open [snapshot (crux/new-snapshot db)]
+                        (->> (for [r reading-ids]
+                               (for [entity-tx (crux/history-ascending db snapshot r)]
+                                 (update entity-tx :crux.db/valid-time #(Date/from (.truncatedTo (.toInstant ^Date %) ChronoUnit/HOURS)))))
+                             (cio/merge-sort (fn [a b]
+                                               (compare (:crux.db/valid-time a) (:crux.db/valid-time b))))
+                             (partition-by :crux.db/valid-time)
+                             (take 12)
+                             (mapv (fn [group]
+                                     (let [battery-levels (sort (mapv (comp :reading/battery-level :crux.db/doc) group))]
+                                       [(:crux.db/valid-time (first group))
+                                        (first battery-levels)
+                                        (last battery-levels)]))))))
+             end-time (System/currentTimeMillis)
+             successful?  (= [[#inst "2016-11-15T12:00:00.000-00:00" 20.0 99.0]
+                              [#inst "2016-11-15T13:00:00.000-00:00" 13.0 100.0]
+                              [#inst "2016-11-15T14:00:00.000-00:00" 9.0 100.0]
+                              [#inst "2016-11-15T15:00:00.000-00:00" 6.0 100.0]
+                              [#inst "2016-11-15T16:00:00.000-00:00" 6.0 100.0]
+                              [#inst "2016-11-15T17:00:00.000-00:00" 6.0 100.0]
+                              [#inst "2016-11-15T18:00:00.000-00:00" 6.0 100.0]
+                              [#inst "2016-11-15T19:00:00.000-00:00" 6.0 100.0]
+                              [#inst "2016-11-15T20:00:00.000-00:00" 6.0 100.0]]
+                             result)]
+         (println (json/write-str {:crux.bench/bench-type ::history-ascending
+                                   ::query-time-ms (- end-time start-time)
+                                   ::successful? successful?})))
+       (catch Exception e
+         (println (json/write-str {:crux.bench/bench-type ::history-ascending
+                                   ::error e}))
+         (throw e))))
+
+
+(defn run-queries [node]
+  (test-10-most-recent-battery-readings-for-charging-devices node)
+  (test-busiest-devices-1-min-avg-whose-battery-level-is-below-33-percent-and-is-not-charging node)
+  (test-min-max-battery-level-per-hour-for-pinto-or-focus-devices node))
+
+(defn bench []
+  (try
+    (with-open [embedded-kafka (ek/start-embedded-kafka
```

Comment:
I think we've got enough in common here to justify extracting this wiring out somewhere common